<h1>Title: Analyzing Users who use Mobile Apps</h1>

<h3>Aim:</h3> This project is about the Mobile app Users of Google Play store and App Store. we will try to analyze how many users use the free apps and paid version of app. This will give us understanding of which apps should we create to engage more number of users.

In [61]:
def openDataSet(file_name,encoding='utf8'):
    opened_file = open(file_name)
    from csv import reader
    read_file = reader(opened_file)
    data = list(read_file)
    return data

In [62]:
applestore_data = openDataSet('AppleStore.csv')

In [63]:
googleplaystore_data = openDataSet('googleplaystore.csv')

In [64]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') 

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [65]:
explore_data(applestore_data,0,5)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']




In [66]:
explore_data(googleplaystore_data,0,5)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']




In [67]:
explore_data(applestore_data,0,1,True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


Number of rows: 7198
Number of columns: 16


In [68]:
explore_data(googleplaystore_data,0,1,True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


Number of rows: 10842
Number of columns: 13


In [69]:
print(googleplaystore_data[10473])

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


In [70]:
del googleplaystore_data[10473]

In [71]:
print(googleplaystore_data[10473])

['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


<h3>Duplicates in dataset:</h3> The dataset contains duplicate entries for various app. This can give us wrong result while performing analysis. This fact can be shown by the example below. let's try finding the duplicate entries for Facebook in google play store data

In [72]:
for app in googleplaystore_data:
    name = app[0]
    if name == 'Facebook':
        print(app[0:10])

['Facebook', 'SOCIAL', '4.1', '78158306', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social']
['Facebook', 'SOCIAL', '4.1', '78128208', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social']


we can also check in google playstore data for the duplicates of Facebook app.

In [73]:
for app in googleplaystore_data:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


<h3>Let's Count the number of Duplicate Entries</h3>

In [74]:
duplicate_apps = []
unique_apps = []

for app in googleplaystore_data:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
        
print('Number of duplicate apps: ',len(duplicate_apps))
print('Number of unique apps: ',len(unique_apps))
print('\n')
print('Example of duplicate apps: ',duplicate_apps[:15])
print('Example of unique apps: ',unique_apps[:15])

Number of duplicate apps:  1181
Number of unique apps:  9660


Example of duplicate apps:  ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']
Example of unique apps:  ['App', 'Photo Editor & Candy Camera & Grid & ScrapBook', 'Coloring book moana', 'U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'Sketch - Draw & Paint', 'Pixel Draw - Number Art Coloring Book', 'Paper flowers instructions', 'Smoke Effect Photo Maker - Smoke Editor', 'Infinite Painter', 'Garden Coloring Book', 'Kids Paint Free - Drawing Fun', 'Text on Photo - Fonteee', 'Name Art Photo Editor - Focus n Filters', 'Tattoo Name On My Photo Editor', 'Mandala Coloring Book']


In [75]:
for app in googleplaystore_data:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


<h4>we can't remove the entries simply because:</h4> For example in above code snippet we can clearly see that 4th row which which corresponds to review is different for each duplicate entries. so The Reaason would be because the number of people given review may had increased. so we can keep the higest humber and remove the others. we have to be always careful in removing all entries. 

<h3>Finding Reviews with highest numbers:</h3>

In [76]:
reviews_max = {}

for val in googleplaystore_data[1:]:
    name = val[0]
    n_reviews = float(val[3])
    if (name in reviews_max) and (reviews_max[name] < n_reviews):
        reviews_max[name] = n_reviews
    if name not in reviews_max:
        reviews_max[name] = n_reviews
        
print(len(reviews_max))
    
    

9659


<h3>Removing the duplicates from the Googleplay Dataset:</h3>

In [77]:
android_clean = []
already_added = []

for val in googleplaystore_data[1:]:
    name = val[0]
    n_reviews = float(val[3])
    if (n_reviews == reviews_max[name]) and (name not in already_added):
        android_clean.append(val)
        already_added.append(name)

In [78]:
print(len(android_clean))

9659


In [79]:
def check_english_old(string):
    for char in string:
        if ord(char) > 127:
            return False
        else:
            return True

In [80]:
check_english_old('Instagram')

True

In [81]:
check_english_old('爱奇艺PPS -《欢乐颂2》电视剧热播')

False

In [82]:
check_english_old('Docs To Go™ Free Office Suite')

True

In [83]:
check_english_old('Instachat 😜')

True

<h3>Check if name of App is english and has more then 3 non English</h3>

In [84]:
def check_english(string):
    non_english = 0
    for char in string:
        if ord(char) > 127:
            non_english += 1
    if non_english > 3:
        return False
    else:
        return True
            

In [85]:
check_english('Docs To Go™ Free Office Suite')

True

In [86]:
check_english('Instachat 😜  😜 😜')

True

In [87]:
check_english('Instachat 😜 😜 😜 😜')

False

In [88]:
check_english('爱奇艺PPS -《欢乐颂2》电视剧热播')

False

<h3>Separating Non-english app from Google play store Dataset</h3>

In [89]:
english_android = []
for row in android_clean:
    name = row[0]
    if check_english(name):
        english_android.append(row)

In [90]:
print(len(english_android))

9614


<h3>Separating Non-english app from Apple store Dataset</h3>

In [91]:
english_ios = []
for row in applestore_data:
    name = row[0]
    if check_english(name):
        english_ios.append(row)

In [92]:
print(len(english_ios))

7198


<h3>Separating Free English app from Google play store Dataset</h3>

In [93]:
print(googleplaystore_data[0:2])

[['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'], ['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']]


In [94]:
print(applestore_data[0:2])

[['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'], ['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']]


In [95]:
free_english_android = []
for row in english_android:
    price = row[7]
    if price == '0':
        free_english_android.append(row)

In [96]:
print(len(free_english_android))

8864


<h3>Separating Free English app from Apple store Dataset</h3>

In [97]:
free_english_apple = []
for row in english_ios:
    price = row[4]
    if price == '0.0':
        free_english_apple.append(row)

In [98]:
print(len(free_english_apple))

4056


<h1>Finding Most common Apps</h1>

<h3>Part One</h3>

As we mentioned in the introduction, our aim is to determine the kinds of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our apps.

To minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:
<ul>
<li>Build a minimal Android version of the app, and add it to Google Play.</li>
<li>If the app has a good response from users, we then develop it further.</li>
<li>If the app is profitable after six months, we also build an iOS version of the app and add it to the App Store.</li>
</ul>
Because our end goal is to add the app on both the App Store and Google Play, we need to find app profiles that are successful on both markets. For instance, a profile that might work well for both markets might be a productivity app that makes use of gamification.

Let's begin the analysis by getting a sense of the most common genres for each market. For this, we'll build a frequency table for the prime_genre column of the App Store data set, and the Genres and Category columns of the Google Play data set.

<h3>Part Two</h3>

In [99]:
def freq_table(dataset, index):
    table = {}
    total = 0
    
    for row in dataset:
        total += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
    percentage_table = {}
    for key in table:
        percentage = (table[key]/total) * 100
        percentage_table[key] = percentage
        
    return percentage_table
        


In [100]:
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [101]:
ios_ferq_prime_genre = display_table(free_english_apple,11)

Games : 55.64595660749507
Entertainment : 8.234714003944774
Photo & Video : 4.117357001972387
Social Networking : 3.5256410256410255
Education : 3.2544378698224854
Shopping : 2.983234714003945
Utilities : 2.687376725838264
Lifestyle : 2.3175542406311638
Finance : 2.0710059171597637
Sports : 1.947731755424063
Health & Fitness : 1.8737672583826428
Music : 1.6518737672583828
Book : 1.6272189349112427
Productivity : 1.5285996055226825
News : 1.4299802761341223
Travel : 1.3806706114398422
Food & Drink : 1.0601577909270217
Weather : 0.7642998027613412
Reference : 0.4930966469428008
Navigation : 0.4930966469428008
Business : 0.4930966469428008
Catalogs : 0.22189349112426035
Medical : 0.19723865877712032


In [102]:
android_freq_category = display_table(free_english_android,1) #category

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

In [103]:
android_freq_genre = display_table(free_english_android,9) #Genre

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

In [104]:
print(android_freq_genre)

None


<h3>Part Three</h3>

<h3>Analyzing prime_genre of App store Data set</h3>


<b>The Most common Genre:</b> 
Games : 55.64595660749507
Entertainment : 8.234714003944774

<b>Other Patterns:</b> Most used apps are of Gaming covering more than 50% then entertainment, photo and video then social networking, education. 

<b>General Impression: </b>Most of the apps which are very famous are entertainment apps not the general impression apps. people are using entertainment apps more then practical purpose apps.

<b>Recommended profile for the App Store market: </b> Gaming profile app is the most recommended app since its mostly used type of app.

<h3>Analyzing Google play store Dataset based on genre and category</h3>

<b>Most common Genres: </b>Most common genres are Tools with 8.44% percent, Entertainment with 6.06%, Education with 5.34%

<b>Other Pattersns: </b><ol><li>More practical purpose genre app like Tools, Entertainment, Business, Productivity apps have more percentage of users using the app.</li>
<li>Category like Family, Game, Tools, Business, Lifestyle has more number of users</li></ol>

<h3>Google play store vs Apple store</h3>

Google play store has more users accesing to practical apps rather than entertainment apps.

In [105]:
some_string = ['FIRST', 'SECOND']
some_integers = [1, 2, 3, 4, 5]

for string in some_string:
    print(string)
    for integer in some_integers:
        print(integer)


FIRST
1
2
3
4
5
SECOND
1
2
3
4
5


<h3>Most Popular Apps by Genre on the App Store</h3>

In [106]:
def freq_table_count(dataset, index):
    table = {}
    total = 0
    
    for row in dataset:
        total += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
    count_genre_table = {}
    for key in table:
        count = table[key]
        count_genre_table[key] = count
        
    return count_genre_table

In [117]:
def display_table_count(dataset, index):
    table = freq_table_count(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [122]:
freq_genre_apple_store = freq_table_count(free_english_apple,11)

In [118]:
freq_genre_apple_store = display_table_count(free_english_apple,11)

Games : 2257
Entertainment : 334
Photo & Video : 167
Social Networking : 143
Education : 132
Shopping : 121
Utilities : 109
Lifestyle : 94
Finance : 84
Sports : 79
Health & Fitness : 76
Music : 67
Book : 66
Productivity : 62
News : 58
Travel : 56
Food & Drink : 43
Weather : 31
Reference : 20
Navigation : 20
Business : 20
Catalogs : 9
Medical : 8


In [123]:
for genre in freq_genre_apple_store:
    total = 0
    len_genre = 0
    for app in free_english_apple:
        genre_app = app[-5]
        if genre_app == genre:
            number_of_user_rating = float(app[5])
            total += number_of_user_rating
            len_genre += 1
    avg_number_of_users = total/len_genre
    print(genre,":",avg_number_of_users)

Music : 56482.02985074627
Food & Drink : 20179.093023255813
Catalogs : 1779.5555555555557
News : 15892.724137931034
Games : 18924.68896765618
Lifestyle : 8978.308510638299
Reference : 67447.9
Health & Fitness : 19952.315789473683
Weather : 47220.93548387097
Productivity : 19053.887096774193
Utilities : 14010.100917431193
Education : 6266.333333333333
Book : 8498.333333333334
Finance : 13522.261904761905
Photo & Video : 27249.892215568863
Shopping : 18746.677685950413
Sports : 20128.974683544304
Social Networking : 53078.195804195806
Business : 6367.8
Travel : 20216.01785714286
Entertainment : 10822.961077844311
Navigation : 25972.05
Medical : 459.75


<b>Music is the recommended app profile for Apple store with "Reference : 67447.9" average users</b>

<h3>Most Popular Apps by Genre on the Google Play Store</h3>

In [121]:
freq_genre_google_play_store = freq_table_count(free_english_android,9)

In [128]:
for category in freq_genre_google_play_store:
    total = 0
    len_category = 0
    for app in free_english_android:
        category_app = app[9]
        if category_app == category:
            number_of_installs = app[5]
            number_of_installs = number_of_installs.replace('+','')
            number_of_installs = number_of_installs.replace(',','')
            total += int(number_of_installs)
            len_category += 1
    avg_install = total/len_category
    print(category,":",avg_install)
    
    

Education;Brain Games : 5333333.333333333
Simulation;Action & Adventure : 4857142.857142857
Casual;Brain Games : 1425916.6666666667
Comics;Creativity : 50000.0
Casual;Creativity : 5333333.333333333
Art & Design;Pretend Play : 500000.0
Strategy;Education : 500000.0
Simulation;Pretend Play : 550000.0
Weather : 5074486.197183099
Maps & Navigation : 4056941.7741935486
Education;Creativity : 2875000.0
Racing : 15910645.681818182
Simulation;Education : 500.0
Entertainment;Action & Adventure : 2333333.3333333335
Role Playing;Pretend Play : 5275000.0
Sports;Action & Adventure : 5050000.0
Entertainment;Music & Video : 6413333.333333333
Health & Fitness;Education : 100000.0
Puzzle;Brain Games : 9280666.666666666
Adventure : 4922785.333333333
Board;Action & Adventure : 3000000.0
Photography : 17840110.40229885
Parenting;Education : 452857.14285714284
Business : 1712290.1474201474
Personalization : 5201482.6122448975
Arcade : 22888365.48780488
Entertainment;Education : 1000000.0
Role Playing;Brain

<h3>Most Popular App on Google Store by Genre is "Communication" with number of users 38456119.17</h3>

<h3>Conclusion:</h3> so Finally by Analyzing both the data of google play store and Apple Store we came up with the conclusion that an app genre that shows potential for being profitable on google play store is of <b>Communication</b> Genre and for App Store is <b>Reference</b> Genre.